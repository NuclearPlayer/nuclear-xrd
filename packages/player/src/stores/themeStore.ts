import { readTextFile } from '@tauri-apps/plugin-fs';
import { create } from 'zustand';

import {
  applyTheme,
  parseThemeJson,
  resolveAdvanced,
  resolveBasic,
  resolveDefaults,
  type ResolvedTheme,
} from '@nuclearplayer/themes';

import { coreSettingsHost } from './settingsStore';

export type ThemeTier = 'basic' | 'advanced';

type ThemeState = {
  resolved?: ResolvedTheme;
  applyFromSettings: () => Promise<void>;
  setTier: (tier: ThemeTier) => Promise<void>;
  setPrimary: (primary: string) => Promise<void>;
  setActiveThemePath: (path: string | undefined) => Promise<void>;
};

const resolveFromSettings = async (): Promise<ResolvedTheme> => {
  const tier = ((await coreSettingsHost.get<string>('theme.tier')) ??
    'basic') as ThemeTier;
  const primary =
    (await coreSettingsHost.get<string>('theme.primary')) ??
    resolveDefaults().light.primary ??
    '';
  const activeThemePath = await coreSettingsHost.get<string | undefined>(
    'theme.activeThemePath',
  );
  if (tier === 'basic') return resolveBasic(primary);
  if (activeThemePath) {
    const text = await readTextFile(activeThemePath);
    const file = parseThemeJson(text);
    return resolveAdvanced(file);
  }
  return resolveDefaults();
};

export const useThemeStore = create<ThemeState>((set) => ({
  resolved: undefined,
  applyFromSettings: async () => {
    const resolved = await resolveFromSettings();
    set({ resolved });
    applyTheme(resolved);
  },
  setTier: async (tier) => {
    await coreSettingsHost.set('theme.tier', tier);
    await useThemeStore.getState().applyFromSettings();
  },
  setPrimary: async (primary) => {
    await coreSettingsHost.set('theme.primary', primary);
    await useThemeStore.getState().applyFromSettings();
  },
  setActiveThemePath: async (path) => {
    await coreSettingsHost.set('theme.activeThemePath', path ?? '');
    await useThemeStore.getState().applyFromSettings();
  },
}));

export const initializeThemeStore = async () => {
  await useThemeStore.getState().applyFromSettings();
};
