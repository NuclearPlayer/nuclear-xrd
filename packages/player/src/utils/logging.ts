import * as Logger from '@tauri-apps/plugin-log';
import isError from 'lodash-es/isError';
import isString from 'lodash-es/isString';
import { toast } from 'sonner';

export const logFsError = async ({
  scope,
  command,
  targetPath,
  error,
  withToast,
  toastMessage,
}: {
  scope: string;
  command: string;
  targetPath: string;
  error: unknown; // Most often Error or string
  withToast?: boolean;
  toastMessage?: string;
}): Promise<void> => {
  const message = error instanceof Error ? error.message : String(error);

  if (withToast) {
    toast.error(toastMessage, { description: message });
  }

  await Logger.error(
    `[${scope}/fs] ${command} failed for ${targetPath}: ${message}`,
  );
};

export const resolveErrorMessage = (error: unknown): string => {
  if (isError(error)) {
    return error.message;
  }
  if (isString(error)) {
    return error;
  }
  return 'Unknown error';
};
