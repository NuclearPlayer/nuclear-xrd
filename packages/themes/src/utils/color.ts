import { converter, formatCss, oklch } from 'culori';

const asOklch = converter('oklch');

export type OklchComponents = { l: number; c: number; h: number };

export const parseOklch = (input: string): OklchComponents | null => {
  const { l, c, h } = oklch(input)!;
  return { l: l as number, c: c as number, h: h as number };
};

export const buildOklch = (components: OklchComponents): string =>
  formatCss({ mode: 'oklch', ...components });

export const replaceHue = (value: string, hue: number): string => {
  const { l, c } = asOklch(value)!;
  return formatCss({ mode: 'oklch', l, c, h: hue });
};
