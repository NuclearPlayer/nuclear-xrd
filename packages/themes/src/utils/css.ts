export const cssVarBlock = (
  selector: string,
  vars: Record<string, string>,
): string => {
  const lines = Object.entries(vars).map(
    ([key, value]) => `  --${key}: ${value};`,
  );
  return `${selector}{
${lines.join('\n')}
}`;
};
