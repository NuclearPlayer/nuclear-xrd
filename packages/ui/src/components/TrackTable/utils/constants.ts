export const DEFAULT_ROW_HEIGHT = 42;
export const DEFAULT_OVERSCAN = 8;
