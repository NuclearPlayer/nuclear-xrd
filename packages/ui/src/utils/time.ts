import { Duration } from 'luxon';

export const formatTime = (totalSeconds: number) => {
  const sign = totalSeconds < 0 ? '-' : '';
  const secs = Math.abs(Math.trunc(totalSeconds));
  const dur = Duration.fromObject({ seconds: secs }).shiftTo(
    'hours',
    'minutes',
    'seconds',
  );
  const hours = dur.hours ?? 0;
  const fmt = hours > 0 ? 'h:mm:ss' : 'm:ss';
  return `${sign}${dur.toFormat(fmt)}`;
};
