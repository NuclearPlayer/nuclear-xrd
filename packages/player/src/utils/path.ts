import { BaseDirectory, exists, mkdir } from '@tauri-apps/plugin-fs';

import { logFsError } from './logging';

export const ensureDir = async (
  dir: string,
  baseDir: BaseDirectory = BaseDirectory.AppData,
) => {
  try {
    const present = await exists(dir, { baseDir });
    if (!present) {
      try {
        await mkdir(dir, {
          recursive: true,
          baseDir,
        });
      } catch (e) {
        await logFsError('ensureDir', 'fs.mkdir', dir, e);
      }
    }
  } catch (e) {
    await logFsError('ensureDir', 'fs.exists', dir, e);
  }
  return dir;
};
