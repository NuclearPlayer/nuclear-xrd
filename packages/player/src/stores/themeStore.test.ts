import '../test/setup';

import { beforeEach, describe, expect, it, vi } from 'vitest';

import { resetInMemoryTauriStore } from '../test/utils/inMemoryTauriStore';
import { joinPath, writeFile } from '../test/utils/testPluginFolder';
import { initializeThemeStore, useThemeStore } from './themeStore';

vi.mock('@tauri-apps/plugin-store', async () => {
  const mod = await import('../test/utils/inMemoryTauriStore');
  return { LazyStore: mod.LazyStore };
});

describe('Theme store', () => {
  beforeEach(() => {
    resetInMemoryTauriStore();
    document.documentElement.removeAttribute('style');
    const existing = document.getElementById('theme-dark-overrides');
    if (existing) existing.remove();
  });

  it('(Snapshot) applies defaults from settings', async () => {
    await initializeThemeStore();
    const light = (
      document.getElementById('theme-light-variables')?.textContent ?? ''
    ).trim();
    const dark = (
      document.getElementById('theme-dark-overrides')?.textContent ?? ''
    ).trim();
    expect({ light, dark }).toMatchSnapshot();
  });

  it('updates on primary change and keeps a single dark style node', async () => {
    await initializeThemeStore();
    const beforeCount = document.querySelectorAll(
      '#theme-dark-overrides',
    ).length;
    expect(beforeCount).toBe(1);
    await useThemeStore.getState().setPrimary('oklch(60% 0.2 210)');
    const afterCount = document.querySelectorAll(
      '#theme-dark-overrides',
    ).length;
    expect(afterCount).toBe(1);
    const light = (
      document.getElementById('theme-light-variables')?.textContent ?? ''
    ).trim();
    expect(light).toContain('oklch(60% 0.2 210)');
  });

  it('(Snapshot) applies advanced theme from JSON file', async () => {
    const themePath = await joinPath('/themes', 'test-theme.json');
    const themeJson = JSON.stringify({
      version: 1,
      name: 'Test Theme',
      variables: {
        background: 'oklch(95% 0.02 120)',
        primary: 'oklch(70% 0.18 160)',
      },
      dark: {
        background: 'oklch(40% 0.08 120)',
      },
    });
    writeFile(themePath, themeJson);
    await initializeThemeStore();
    await useThemeStore.getState().setTier('advanced');
    await useThemeStore.getState().setActiveThemePath(themePath);
    const light = document.documentElement.getAttribute('style') ?? '';
    const dark = (
      document.getElementById('theme-dark-overrides')?.textContent ?? ''
    ).trim();
    expect({ light, dark }).toMatchSnapshot();
  });
});
