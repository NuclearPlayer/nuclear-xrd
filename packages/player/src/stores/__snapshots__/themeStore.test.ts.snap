// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Theme store > (Snapshot) applies advanced theme from JSON file 1`] = `
{
  "dark": "[data-theme='dark']{
  --background: oklch(40% 0.08 120);
  --background-secondary: oklch(39.49% 0.031 277.49);
  --background-input: oklch(0% 0 0);
  --foreground: oklch(0.9249 0 0);
  --foreground-secondary: oklch(0.9249 0 0);
  --foreground-input: oklch(100% 0 0);
}",
  "light": "--background: oklch(95% 0.02 120); --background-secondary: oklch(100% 0 0); --background-input: oklch(100% 0 0); --foreground: oklch(0% 0 0); --foreground-secondary: oklch(0.3 0 0); --foreground-input: oklch(0% 0 0); --primary: oklch(70% 0.18 160); --border: oklch(0% 0 0); --ring: oklch(100% 0 0); --accent-green: oklch(79.05% 0.209 147.58); --accent-yellow: oklch(95.53% 0.134 112.76); --accent-purple: oklch(74.2% 0.149 301.88); --accent-blue: oklch(55.98% 0.08 270.09); --accent-orange: oklch(83.39% 0.124 66.56); --accent-cyan: oklch(88.26% 0.093 212.85); --accent-red: oklch(68.22% 0.206 24.43); --shadow-color: oklch(0% 0 0); --shadow-x: 2px; --shadow-y: 2px; --shadow-blur: 0px; --radius: 8px; --radius-sm: 4px; --radius-lg: 12px; --font-family: 'DM Sans', system-ui, -apple-system, sans-serif;",
}
`;

exports[`Theme store > (Snapshot) applies defaults from settings 1`] = `
{
  "dark": "[data-theme='dark']{
  --background: oklch(0.4408 0.0838 341.75);
  --background-secondary: oklch(39.49% 0.031 277.49);
  --background-input: oklch(0% 0 0);
  --foreground: oklch(0.9249 0 0);
  --foreground-secondary: oklch(0.9249 0 0);
  --foreground-input: oklch(100% 0 0);
}",
  "light": "--background: oklch(0.9491 0.023 341.75); --background-secondary: oklch(100% 0 0); --background-input: oklch(100% 0 0); --foreground: oklch(0% 0 0); --foreground-secondary: oklch(0.3 0 0); --foreground-input: oklch(0% 0 0); --primary: oklch(76.91% 0.173 341.75); --border: oklch(0% 0 0); --ring: oklch(100% 0 0); --accent-green: oklch(79.05% 0.209 147.58); --accent-yellow: oklch(95.53% 0.134 112.76); --accent-purple: oklch(74.2% 0.149 301.88); --accent-blue: oklch(55.98% 0.08 270.09); --accent-orange: oklch(83.39% 0.124 66.56); --accent-cyan: oklch(88.26% 0.093 212.85); --accent-red: oklch(68.22% 0.206 24.43); --shadow-color: oklch(0% 0 0); --shadow-x: 2px; --shadow-y: 2px; --shadow-blur: 0px; --radius: 8px; --radius-sm: 4px; --radius-lg: 12px; --font-family: 'DM Sans', system-ui, -apple-system, sans-serif;",
}
`;
